src/main/java/io/cryostat/configuration/CredentialsManager.java

```diff
@@ -134,70 +159,29 @@ public static String targetIdToMatchExpression(String targetId) {
         return String.format("target.connectUrl == \"%s\"", targetId);
     }
 
-    public void load() throws IOException {
-        this.nextId =
-                this.fs.listDirectoryChildren(credentialsDir).stream()
-                        .peek(n -> logger.trace("Credentials file: {}", n))
-                        .map(credentialsDir::resolve)
-                        .map(
-                                path -> {
-                                    try {
-                                        String filename = path.getFileName().toString();
-                                        return Integer.parseInt(filename);
-                                    } catch (NumberFormatException nfe) {
-                                        logger.error(nfe);
-                                        try {
-                                            fs.deleteIfExists(path);
-                                        } catch (IOException ioe) {
-                                            logger.error(ioe);
-                                        }
-                                        return 0;
-                                    }
-                                })
-                        .reduce(Math::max)
-                        .orElse(0);
-    }
-
     public int addCredentials(String matchExpression, Credentials credentials)
-            throws IOException, MatchExpressionValidationException {
+            throws MatchExpressionValidationException {
         matchExpressionValidator.validate(matchExpression);
-        Path destination = getPersistedPath(nextId);
-        fs.writeString(
-                destination,
-                gson.toJson(new StoredCredentials(matchExpression, credentials)),
-                StandardOpenOption.WRITE,
-                StandardOpenOption.CREATE,
-                StandardOpenOption.TRUNCATE_EXISTING);
-        fs.setPosixFilePermissions(
-                destination,
-                Set.of(PosixFilePermission.OWNER_READ, PosixFilePermission.OWNER_WRITE));
+        StoredCredentials saved = dao.save(new StoredCredentials(matchExpression, credentials));
         emit(CredentialsEvent.ADDED, matchExpression);
-        return nextId++;
+        return saved.getId();
     }
 
-    public int removeCredentials(String matchExpression)
-            throws IOException, MatchExpressionValidationException {
+    public int removeCredentials(String matchExpression) throws MatchExpressionValidationException {
+        // TODO refactor this to do a proper query on the DAO
```

Comment:
Is this still a `TODO`?

Reply:
Yes, I tried to solve this before but ran into some issue that I don't fully remember now. I *think* the problem was that doing the removal with a delete query in the DAO wouldn't return back the ID, so the return value from this method would need to change and that would change the behaviour of or break the original TargetCredentialsDeleteHandler.

The target-specific API handlers should be deprecated since they are now superseded by the matchExpression credentials API, so once those are removed we can actually just remove this method altogether and skip the TODO.

Reply:
Right, make senses! Thanks for explaining :D